src/librbd/api/Io.cc

```diff
@@ -0,0 +1,372 @@
+                 << "len = " << len << dendl;
+
+  image_ctx.image_lock.lock_shared();
+  int r = clip_io(util::get_image_ctx(&image_ctx), off, &len);
+  image_ctx.image_lock.unlock_shared();
+  if (r < 0) {
+    lderr(cct) << "invalid IO request: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  C_SaferCond ctx;
+  auto aio_comp = io::AioCompletion::create(&ctx);
+  aio_discard(image_ctx, aio_comp, off, len, discard_granularity_bytes, false);
+
+  r = ctx.wait();
+  if (r < 0) {
+    return r;
+  }
+  return len;
+}
+
+template <typename I>
+ssize_t Io<I>::write_same(
+    I &image_ctx, uint64_t off, uint64_t len, bufferlist &&bl, int op_flags) {
+  auto cct = image_ctx.cct;
+  ldout(cct, 20) << "ictx=" << &image_ctx << ", off=" << off << ", "
+                 << "len = " << len << ", data_len " << bl.length() << dendl;
+
+  image_ctx.image_lock.lock_shared();
+  int r = clip_io(util::get_image_ctx(&image_ctx), off, &len);
+  image_ctx.image_lock.unlock_shared();
+  if (r < 0) {
+    lderr(cct) << "invalid IO request: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  C_SaferCond ctx;
+  auto aio_comp = io::AioCompletion::create(&ctx);
+  aio_write_same(image_ctx, aio_comp, off, len, std::move(bl), op_flags, false);
+
+  r = ctx.wait();
+  if (r < 0) {
+    return r;
+  }
+  return len;
+}
+
+template <typename I>
+ssize_t Io<I>::compare_and_write(
+    I &image_ctx, uint64_t off, uint64_t len, bufferlist &&cmp_bl,
+    bufferlist &&bl, uint64_t *mismatch_off, int op_flags) {
+  auto cct = image_ctx.cct;
+  ldout(cct, 20) << "compare_and_write ictx=" << &image_ctx << ", off="
+                 << off << ", " << "len = " << len << dendl;
+
+  image_ctx.image_lock.lock_shared();
+  int r = clip_io(util::get_image_ctx(&image_ctx), off, &len);
+  image_ctx.image_lock.unlock_shared();
+  if (r < 0) {
+    lderr(cct) << "invalid IO request: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  C_SaferCond ctx;
+  auto aio_comp = io::AioCompletion::create(&ctx);
+  aio_compare_and_write(image_ctx, aio_comp, off, len, std::move(cmp_bl),
+                        std::move(bl), mismatch_off, op_flags, false);
+
+  r = ctx.wait();
+  if (r < 0) {
+    return r;
+  }
+  return len;
+}
+
+template <typename I>
+int Io<I>::flush(I &image_ctx) {
+  auto cct = image_ctx.cct;
+  ldout(cct, 20) << "ictx=" << &image_ctx << dendl;
+
+  C_SaferCond ctx;
+  auto aio_comp = io::AioCompletion::create(&ctx);
+  aio_flush(image_ctx, aio_comp, false);
+
+  int r = ctx.wait();
+  if (r < 0) {
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+void Io<I>::aio_read(I &image_ctx, io::AioCompletion *aio_comp, uint64_t off,
+                     uint64_t len, io::ReadResult &&read_result, int op_flags,
+                     bool native_async) {
+  auto cct = image_ctx.cct;
+  FUNCTRACE(cct);
+  ZTracer::Trace trace;
+  if (image_ctx.blkin_trace_all) {
+    trace.init("wq: read", &image_ctx.trace_endpoint);
+    trace.event("start");
```

Comment:
The event is called "start" hear, but in all other `aio_*` functions it is called "init". Is it intentional?

Reply:
No, copy-and-pasted from `ImageRequestWQ`. I've cleaned it up, thx.